### PR TITLE
Xbox controller

### PR DESCRIPTION
**Pull request by russell-taylor**
This adds a directory that includes an implementation for a radial-distortion-correction routine that let the user change parameters using an Xbox controller while viewing a heads-up display showing a grid and information about the parameters.

### CODE REVIEW

**Comment by russell-taylor** [comment]
Adds a utility to let you adjust parameters for a radially-symmetric distortion parameters while viewing their impact.  Uses an xbox controller to adjust them.

File: RenderManager_xbox/DistortionCorrectRenderManager.cpp [hunk 1]
```diff
@@ -0,0 +1,760 @@
...
+} XY;
+
+static std::vector<float> params;  //< Distortion parameters
+static int activeParam = 0;  //< Which parameter are we adjusting?
+
+static std::string osvrGetString(OSVR_ClientContext context, const std::string& path)
+{
+  size_t len;
+  if (osvrClientGetStringParameterLength(context, path.c_str(), &len) == OSVR_RETURN_FAILURE) {
+    std::string msg = std::string("Couldn't get osvr string length for path ") + path;
```
Comment (mars979) [inline RenderManager_xbox/DistortionCorrectRenderManager.cpp]: Nothing wrong with this, but you can avoid creating msg var and just use `std::cerr << "Message " << path << std::endl`.

File: RenderManager_xbox/DistortionCorrectRenderManager.cpp [hunk 2]
```diff
@@ -0,0 +1,760 @@
...
+
+  //struct TempBuffer {
+  //    char *buffer;
+  //    TempBuffer(size_t len) { buffer = new char[len + 1]; }
+  //    ~TempBuffer() { delete[] buffer; }
+  //} tempBuffer(len);
+  std::vector<char> tempBuffer(len + 1);
+
+  if (osvrClientGetStringParameter(context, path.c_str(), tempBuffer.data(), len + 1) == OSVR_RETURN_FAILURE) {
+    std::string msg = std::string("Couldn't get osvr string buffer for path ") + path;
```
Comment (mars979) [inline RenderManager_xbox/DistortionCorrectRenderManager.cpp]: Same here as above

File: RenderManager_xbox/DistortionCorrectRenderManager.cpp [hunk 3]
```diff
@@ -0,0 +1,760 @@
...
+    reinterpret_cast<OSVRDisplayConfiguration *>(userdata);
+
+  if (report->state == 1) {
+
+    // Create a new set of distortion parameters that has the
+    // specified parameters, but using the center of projection
+    // from the read-in values.
+    std::cout << "XXX New parameters: ";
+    for (size_t i = 0; i < params.size(); i++) {
+      std::cout << params[i] << " ";
```
Comment (mars979) [inline RenderManager_xbox/DistortionCorrectRenderManager.cpp]: The comment says that it creates new set of params, but this block just prints them out. Is this just left when you were debugging?

File: RenderManager_xbox/DistortionCorrectRenderManager.cpp [hunk 4]
```diff
@@ -0,0 +1,760 @@
...
+}
+
+void resetParams(void *userdata, const OSVR_TimeValue *timestamp,
+  const OSVR_ButtonReport *report)
+{
+  if (report->state == 1) {
+    for (size_t i = 0; i < params.size(); i++) {
+      params[i] = 0;
+    }
+    params[1] = 1;
```
Comment (mars979) [inline RenderManager_xbox/DistortionCorrectRenderManager.cpp]: Are you setting 2nd parameter to 1 on purpose here after resetting to 0 before?

File: RenderManager_xbox/DistortionCorrectRenderManager.cpp [hunk 5]
```diff
@@ -0,0 +1,760 @@
...
+    if (!render->RegisterRenderBuffers(colorBuffers)) {
+      std::cerr << "RegisterRenderBuffers() returned false, cannot continue" << std::endl;
+      quit = true;
+    }
+
+    // Generate the vector of sphere locations we're going to use to control
+    // distortion correction.
+    std::vector<XY> spheres;
+    const size_t numSpheres = 1;
+    const double sphereSpace = 1.0 / (numSpheres);
```
Comment (mars979) [inline RenderManager_xbox/DistortionCorrectRenderManager.cpp]: looks like you are just dividing here by 1 so sphereSpace still remains 1.0 (unless you will be changing numSpheres in future)

File: RenderManager_xbox/DistortionCorrectRenderManager.cpp [hunk 6]
```diff
@@ -0,0 +1,760 @@
...
+      sphere.x = i * sphereSpace;
+      sphere.y = 0;
+      spheres.push_back(sphere);
+    }
+
+    // Initialize the distortion parameters, six parameters but only
+    // the linear one is nonzero.
+
+    /*
+    params.push_back(0);
```
Comment (mars979) [inline RenderManager_xbox/DistortionCorrectRenderManager.cpp]: Could probably remove commented out params

File: RenderManager_xbox/DistortionCorrectRenderManager.cpp (hunk 1 above)
Comment (russell-taylor) [inline RenderManager_xbox/DistortionCorrectRenderManager.cpp]: Replaced.

File: RenderManager_xbox/DistortionCorrectRenderManager.cpp (hunk 2 above)
Comment (russell-taylor) [inline RenderManager_xbox/DistortionCorrectRenderManager.cpp]: Except I then need to pass the message as a throw

File: RenderManager_xbox/DistortionCorrectRenderManager.cpp (hunk 1 above)
Comment (russell-taylor) [inline RenderManager_xbox/DistortionCorrectRenderManager.cpp]: Unreplaced... I then throw using the msg.

File: RenderManager_xbox/DistortionCorrectRenderManager.cpp (hunk 3 above)
Comment (russell-taylor) [inline RenderManager_xbox/DistortionCorrectRenderManager.cpp]: That comment is for the whole section of code following.  Removed the debug print.

File: RenderManager_xbox/DistortionCorrectRenderManager.cpp (hunk 4 above)
Comment (russell-taylor) [inline RenderManager_xbox/DistortionCorrectRenderManager.cpp]: Yes, this produces [ 0, 1, 0+ ].  Added comment to make this clear.

File: RenderManager_xbox/DistortionCorrectRenderManager.cpp (hunk 1 above)
Comment (mars979) [inline RenderManager_xbox/DistortionCorrectRenderManager.cpp]: Gotcha, sorry, somehow missed the throw.

File: RenderManager_xbox/DistortionCorrectRenderManager.cpp (hunk 5 above)
Comment (russell-taylor) [inline RenderManager_xbox/DistortionCorrectRenderManager.cpp]: This was made for a general number of spheres, but then I ended up using one.

File: RenderManager_xbox/DistortionCorrectRenderManager.cpp (hunk 6 above)
Comment (russell-taylor) [inline RenderManager_xbox/DistortionCorrectRenderManager.cpp]: Done.

**Comment by russell-taylor** [comment]
I think I fixed the issues.  Go ahead and merge if you agree.

**Pull request merged**